packages/@jsii/kernel/src/disk-cache/disk-cache.ts

```diff
@@ -0,0 +1,244 @@
+import {
+  existsSync,
+  mkdirSync,
+  readdirSync,
+  readFileSync,
+  realpathSync,
+  rmdirSync,
+  rmSync,
+  statSync,
+  utimesSync,
+  writeFileSync,
+} from 'fs';
+import { lockSync, unlockSync } from 'lockfile';
+import { dirname, join } from 'path';
+
+import { digestFile } from './digest-file';
+
+const MARKER_FILE_NAME = '.jsii-runtime-package-cache';
+
+const ONE_DAY_IN_MS = 86_400_000;
+const PRUNE_AFTER_MILLISECONDS = process.env.JSII_RUNTIME_PACKAGE_CACHE_TTL
+  ? parseInt(process.env.JSII_RUNTIME_PACKAGE_CACHE_TTL, 10) * ONE_DAY_IN_MS
+  : 30 * ONE_DAY_IN_MS;
+
+export class DiskCache {
+  private static readonly CACHE = new Map<string, DiskCache>();
+
+  public static inDirectory(path: string): DiskCache {
+    const didCreate = mkdirSync(path, { recursive: true }) != null;
+    if (didCreate && process.platform === 'darwin') {
+      // Mark the directories for no iCloud sync, no Spotlight indexing, no TimeMachine backup
+      // @see https://michaelbach.de/2019/03/19/MacOS-nosync-noindex-nobackup.html
+      writeFileSync(join(path, '.nobackup'), '');
+      writeFileSync(join(path, '.noindex'), '');
+      writeFileSync(join(path, '.nosync'), '');
+    }
+
+    path = realpathSync(path);
+    if (!this.CACHE.has(path)) {
+      this.CACHE.set(path, new DiskCache(path));
+    }
+    return this.CACHE.get(path)!;
+  }
+
+  readonly #root: string;
+
+  private constructor(root: string) {
+    this.#root = root;
+    process.once('beforeExit', () => this.pruneExpiredEntries());
+  }
+
+  public entry(...key: readonly string[]): Entry {
+    if (key.length === 0) {
+      throw new Error(`Cache entry key must contain at least 1 element!`);
+    }
+
+    return new Entry(
+      join(
+        this.#root,
+        ...key.flatMap((s) =>
+          s
+            .replace(/[^@a-z0-9_.\\/-]+/g, '_')
+            .split(/[\\/]+/)
+            .map((ss) => {
+              if (ss === '..') {
+                throw new Error(
+                  `A cache entry key cannot contain a '..' path segment! (${s})`,
+                );
+              }
+              return ss;
+            }),
+        ),
+      ),
+    );
+  }
+
+  public entryFor(path: string, ...comments: readonly string[]): Entry {
+    const rawDigest = digestFile(path, ...comments);
+    return this.entry(...comments, rawDigest.toString('hex'));
+  }
+
+  public pruneExpiredEntries() {
+    const cutOff = new Date(Date.now() - PRUNE_AFTER_MILLISECONDS);
+    for (const entry of this.entries()) {
```

Comment:
[Q] Can this lead to a strange part-old part to be renewed situation and would it matter? Are there dependencies between the cached items?

Reply:
Cached items are literally "the extracted content of a tarball". They technically don't depend on each other - the "installs" of these are copies/hard-links in another directory (temporary). If/when a directory is removed from cache, the corresponding installs remain "valid" (since they were hard linked into place).

The only consequence of expiring something from this cache is that it's going to need to be re-extracted next time...